integration/log_integration_test.go

```diff
@@ -0,0 +1,333 @@
+	// Have to work with strings because slices can't be map keys. Sigh.
+	leafDataPresenceMap := make(map[string]bool)
+
+	for l := int64(0); l < params.leafCount; l++ {
+		leafDataPresenceMap[fmt.Sprintf("Leaf %d", l + params.startLeaf)] = true
+	}
+
+	for currentLeaf < params.leafCount {
+		hasher := merkle.NewRFC6962TreeHasher(trillian.NewSHA256())
+
+		// We have to allow for the last batch potentially being a short one
+		numLeaves := params.leafCount - currentLeaf
+
+		if numLeaves > int64(params.readBatchSize) {
+			numLeaves = int64(params.readBatchSize)
+		}
+
+		glog.Infof("Reading %d leaves from %d ...", numLeaves, currentLeaf + params.startLeaf)
+		req := makeGetLeavesByIndexRequest(logID, currentLeaf + params.startLeaf, numLeaves)
+		ctx, cancelFunc := getRPCDeadlineContext()
+		response, err := client.GetLeavesByIndex(ctx, req)
+		cancelFunc()
+
+		if err != nil {
+			return nil, err
+		}
+
+		if response.Status == nil || response.Status.StatusCode != trillian.TrillianApiStatusCode_OK {
+			return nil, fmt.Errorf("read leaves failed: %s %d", response.Status.Description, response.Status.StatusCode)
+		}
+
+		// Check we got the right leaf count
+		if int64(len(response.Leaves)) != numLeaves {
+			return nil, fmt.Errorf("expected %d leaves but we only read %d", numLeaves, len(response.Leaves))
+		}
+
+		// Check the leaf contents make sense. Can't rely on exact ordering as queue timestamps will be
+		// close between batches and identical within batches.
+		for l := 0; l < len(response.Leaves); l++ {
+			// Check for duplicate leaf index in response data - should not happen
+			if _, ok := leafMap[response.Leaves[l].LeafIndex]; ok {
+				return nil, fmt.Errorf("got duplicate leaf sequence number: %d", response.Leaves[l].LeafIndex)
+			}
+
+			leafMap[response.Leaves[l].LeafIndex] = response.Leaves[l]
+
+			// Test for having seen duplicate leaf data - it should all be distinct
+			_, ok := leafDataPresenceMap[string(response.Leaves[l].LeafValue)]
+
+			if !ok {
+				return nil, fmt.Errorf("leaf data duplicated for leaf: %v", response.Leaves[l])
+			}
+
+			delete(leafDataPresenceMap, string(response.Leaves[l].LeafValue))
+
+			hash := hasher.HashLeaf(response.Leaves[l].LeafValue)
+
+			if !bytes.Equal(hash[:], response.Leaves[l].MerkleLeafHash) {
+				return nil, fmt.Errorf("leaf hash mismatch expected: %s, got: %s", base64.StdEncoding.EncodeToString(hash[:]), base64.StdEncoding.EncodeToString(response.Leaves[l].MerkleLeafHash))
+			}
+		}
+
+		currentLeaf += int64(params.readBatchSize)
+	}
+
+	// By this point we expect to have seen all the leaves so there should be nothing in the map
+	if len(leafDataPresenceMap) != 0 {
+		return nil, fmt.Errorf("missing leaves from data read back: %v", leafDataPresenceMap)
+	}
+
+	return leafMap, nil
+}
+
+func checkLogSTHConsistency(logID trillian.LogID, tree *merkle.InMemoryMerkleTree, client trillian.TrillianLogClient, params testParameters) error {
```

Comment:
I'd rename this method, if I were skimming I'd assume it's fetching and checking a consistency proof.